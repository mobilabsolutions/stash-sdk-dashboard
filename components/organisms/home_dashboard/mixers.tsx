import React from 'react'
import { useState, useEffect } from 'react'
import {
  useKeyPerformance,
  useDayActivity,
  useNotifications
} from '../../../hooks'
import { ReactElementLike } from 'prop-types'
import {
  KeyPerformance,
  TodaysActivity,
  Notifications,
  StaticNotifications
} from '../../../types'
import moment from 'moment'

//===========> Key Performance
interface KPProps {
  children: (p: KeyPerformance) => ReactElementLike
  liveData: KeyPerformance
}

export function KPMixer(props: KPProps) {
  const { children, liveData } = props

  const { kp, allowStream } = useKeyPerformance()
  const [state, setState] = useState({
    nrOfChargebacks: 0,
    nrOfRefundedTransactions: 0,
    nrOfTransactions: 0,
    salesVolume: 0,
    currencyId: 'EUR'
  })
  useEffect(() => {
    allowStream && setState(kp)
  }, [allowStream])
  useEffect(() => {
    allowStream &&
      liveData &&
      setState(prev => ({
        nrOfChargebacks: liveData.nrOfChargebacks + prev.nrOfChargebacks,
        nrOfRefundedTransactions:
          liveData.nrOfRefundedTransactions + prev.nrOfRefundedTransactions,
        nrOfTransactions: liveData.nrOfTransactions + prev.nrOfTransactions,
        salesVolume: liveData.salesVolume + prev.salesVolume,
        currencyId: prev.currencyId
      }))
  }, [liveData])
  return <>{children(state)}</>
}

//===========> Today's Activity
interface TAProps {
  children: (p: { data: MixedResult[] }) => ReactElementLike
  selectedDay: moment.Moment
  liveData: TodaysActivity
}

const timeReg = /^(?:(?:([01]?\d|2[0-3]):)?([0-5]?\d):)?([0-5]?\d)$/

interface MixedResult {
  today: number
  selectedDay: number
  time: string
}

const mapToPlot = (key: string) => (p: TodaysActivity) => {
  return {
    time: Number(p.time),
    [key]: p.amount / 100
  }
}

function mixDataForPlot(
  live: TodaysActivity[] = [],
  staticD: TodaysActivity[] = []
): MixedResult[] {
  const plotedSelectedDay = staticD.map(mapToPlot('selectedDay'))
  const plotedToday = live.map(mapToPlot('today'))
  const mappedData = plotedToday
    .concat(plotedSelectedDay)
    .reduce((acum, curr) => {
      const existing = acum[curr.time] || {}
      return {
        ...acum,
        [curr.time]: {
          ...existing,
          ...curr
        }
      }
    }, {})
  const flattedData = Object.entries(mappedData).map(
    ([_key, entry]) => entry as MixedResult
  )
  return flattedData
}

export function TAMixer(props: TAProps) {
  const { selectedDay, children, liveData } = props
  const { activities: selectedDayActivity } = useDayActivity(selectedDay)
  const { activities: todayActivities } = useDayActivity(moment())
  const [live, setLive] = useState<TodaysActivity[]>([])
  useEffect(() => {
    const nowHour = moment().hour()
    setLive(todayActivities.filter(act => Number(act.time) <= nowHour))
  }, [todayActivities])

  useEffect(() => {
    if (liveData) {
      const lastAct = live.length
        ? live[live.length - 1]
        : { amount: 0, time: moment().hour() }
      const [, ldHr] = timeReg.exec(liveData.time)

      const shouldPush = Number(ldHr) > Number(lastAct.time)
      if (shouldPush) {
        const modifyedLast = { ...liveData, time: ldHr }
        setLive(prev => [...prev, modifyedLast])
      } else {
        setLive(prev =>
          prev.map((act, i) =>
            i === prev.length - 1
              ? { ...act, amount: act.amount + liveData.amount }
              : act
          )
        )
      }
    }
    // !!liveData && setLive(prev => [...prev, liveData])
  }, [liveData])

  return <>{children({ data: mixDataForPlot(live, selectedDayActivity) })}</>
}

//===========> Notifications

interface NProps {
  children: (props: StaticNotifications) => ReactElementLike
  liveData: Notifications
}

const mapDays = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
]

export function NotificationMixer(props: NProps) {
  const { data } = useNotifications()
  const { liveData, children } = props
  const [state, setState] = useState<StaticNotifications>(data)
  const moment_en = moment()
  moment_en.locale('en')
  const _day = mapDays[moment_en.weekday()] // Using this approach in order to get the same weekday if diferent locale
  useEffect(() => {
    setState(data)
  }, [data])
  useEffect(() => {
    if (liveData) {
      setState({
        notifications: liveData.notification
          ? [liveData.notification, ...state.notifications]
          : state.notifications,
        transactions: [
          ...state.transactions.map(({ day, nrOfTransactions }) =>
            day === _day
              ? {
                  day,
                  nrOfTransactions:
                    nrOfTransactions + (liveData.nrOfTransactions || 0)
                }
              : { day, nrOfTransactions }
          )
        ]
      })
    }
  }, [liveData])
  return <>{children(state)}</>
}
