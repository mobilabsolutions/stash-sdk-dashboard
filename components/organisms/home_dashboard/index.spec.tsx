jest.mock('../../../hooks/use_key_performance', () => () => {
  //mock hook to avoid fetch calls
  return {
    kp: {
      nrOfChargebacks: 10,
      nrOfRefundedTransactions: 32,
      nrOfTransactions: 43,
      salesVolume: 54,
      currencyId: 'EUR'
    },
    allowStream: true,
    loading: false,
    error: null
  }
})

import { KPMixer } from './mixers'
import KeyPerformance from '../key_performance'
import { deepRender } from '../../../test_utils'

const withMix = (Mixer, Cmp) => ({ liveData }) => (
  <Mixer liveData={liveData}>{(props: any) => <Cmp {...props} />}</Mixer>
)

describe('Mixer for "live" and "static" data', () => {
  test('should work for "KeyPerformance"', () => {
    const Mixed = withMix(KPMixer, KeyPerformance)
    const { getAllByTestId } = deepRender(Mixed, {
      liveData: {
        nrOfChargebacks: 12,
        nrOfRefundedTransactions: 0,
        nrOfTransactions: 2,
        salesVolume: 3,
        currencyId: 'EUR'
      }
    })

    expect(getAllByTestId('card-title')).toMatchSnapshot()
  })
  //... Will continue
})
