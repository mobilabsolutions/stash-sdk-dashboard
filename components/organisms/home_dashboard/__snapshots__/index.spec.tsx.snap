// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mixer for "live" and "static" data should work for "KeyPerformance" 1`] = `
Array [
  <div
    class="right"
    data-testid="card-title"
  >
    <h1
      class="h1-sc-1ivb2ca-0 performance_card__Title-sc-1b7j6s7-1 cIKajw"
      style="color: rgb(96, 157, 246);"
    >
      â‚¬0.57
    </h1>
    <span>
      Sales Volume
    </span>
  </div>,
  <div
    class="right"
    data-testid="card-title"
  >
    <h1
      class="h1-sc-1ivb2ca-0 performance_card__Title-sc-1b7j6s7-1 cIKajw"
      style="color: rgb(79, 89, 149);"
    >
      45
    </h1>
    <span>
      Transactions
    </span>
  </div>,
  <div
    class="right"
    data-testid="card-title"
  >
    <h1
      class="h1-sc-1ivb2ca-0 performance_card__Title-sc-1b7j6s7-1 cIKajw"
      style="color: rgb(4, 139, 133);"
    >
      32
    </h1>
    <span>
      Refunded Transactions
    </span>
  </div>,
  <div
    class="right"
    data-testid="card-title"
  >
    <h1
      class="h1-sc-1ivb2ca-0 performance_card__Title-sc-1b7j6s7-1 cIKajw"
      style="color: rgb(242, 121, 105);"
    >
      22
    </h1>
    <span>
      Chargebacks
    </span>
  </div>,
]
`;

exports[`Mixer for "live" and "static" data should work for "Todays Activity": render al series 1`] = `
<div>
  <div
    data-testid="0"
  >
    <span
      data-testid="today-0"
    >
      452.54
    </span>
    <span
      data-testid="selected-0"
    >
      45234.52
    </span>
  </div>
  <div
    data-testid="1"
  >
    <span
      data-testid="today-1"
    >
      1231.34
    </span>
    <span
      data-testid="selected-1"
    >
      453.53
    </span>
  </div>
  <div
    data-testid="2"
  >
    <span
      data-testid="today-2"
    >
      25.62
    </span>
    <span
      data-testid="selected-2"
    >
      4525.42
    </span>
  </div>
  <div
    data-testid="3"
  >
    <span
      data-testid="today-3"
    >
      52.45
    </span>
    <span
      data-testid="selected-3"
    >
      312.41
    </span>
  </div>
  <div
    data-testid="4"
  >
    <span
      data-testid="today-4"
    >
      6373.53
    </span>
    <span
      data-testid="selected-4"
    >
      121.41
    </span>
  </div>
  <div
    data-testid="5"
  >
    <span
      data-testid="today-5"
    >
      88984.56
    </span>
    <span
      data-testid="selected-5"
    >
      453.52
    </span>
  </div>
  <div
    data-testid="6"
  >
    <span
      data-testid="today-6"
    >
      453456.79
    </span>
    <span
      data-testid="selected-6"
    >
      24556.56
    </span>
  </div>
  <div
    data-testid="7"
  >
    <span
      data-testid="today-7"
    >
      3434.51
    </span>
    <span
      data-testid="selected-7"
    >
      234.52
    </span>
  </div>
  <div
    data-testid="8"
  >
    <span
      data-testid="today-8"
    >
      18
    </span>
    <span
      data-testid="selected-8"
    >
      5432.54
    </span>
  </div>
  <div
    data-testid="9"
  >
    <span
      data-testid="today-9"
    >
      84
    </span>
    <span
      data-testid="selected-9"
    >
      189.19
    </span>
  </div>
  <div
    data-testid="10"
  >
    <span
      data-testid="today-10"
    >
      218.82
    </span>
    <span
      data-testid="selected-10"
    >
      401.49
    </span>
  </div>
  <div
    data-testid="11"
  >
    <span
      data-testid="today-11"
    />
    <span
      data-testid="selected-11"
    >
      533.74
    </span>
  </div>
  <div
    data-testid="12"
  >
    <span
      data-testid="today-12"
    />
    <span
      data-testid="selected-12"
    >
      48
    </span>
  </div>
  <div
    data-testid="13"
  >
    <span
      data-testid="today-13"
    />
    <span
      data-testid="selected-13"
    >
      163.22
    </span>
  </div>
  <div
    data-testid="14"
  >
    <span
      data-testid="today-14"
    />
    <span
      data-testid="selected-14"
    >
      97.44
    </span>
  </div>
  <div
    data-testid="15"
  >
    <span
      data-testid="today-15"
    />
    <span
      data-testid="selected-15"
    >
      77.54
    </span>
  </div>
  <div
    data-testid="16"
  >
    <span
      data-testid="today-16"
    />
    <span
      data-testid="selected-16"
    >
      345.22
    </span>
  </div>
  <div
    data-testid="17"
  >
    <span
      data-testid="today-17"
    />
    <span
      data-testid="selected-17"
    >
      235624.56
    </span>
  </div>
  <div
    data-testid="18"
  >
    <span
      data-testid="today-18"
    />
    <span
      data-testid="selected-18"
    >
      54.14
    </span>
  </div>
  <div
    data-testid="19"
  >
    <span
      data-testid="today-19"
    />
    <span
      data-testid="selected-19"
    >
      35134.32
    </span>
  </div>
  <div
    data-testid="20"
  >
    <span
      data-testid="today-20"
    />
    <span
      data-testid="selected-20"
    >
      0
    </span>
  </div>
  <div
    data-testid="21"
  >
    <span
      data-testid="today-21"
    />
    <span
      data-testid="selected-21"
    >
      0
    </span>
  </div>
  <div
    data-testid="22"
  >
    <span
      data-testid="today-22"
    />
    <span
      data-testid="selected-22"
    >
      0
    </span>
  </div>
  <div
    data-testid="23"
  >
    <span
      data-testid="today-23"
    />
    <span
      data-testid="selected-23"
    >
      0
    </span>
  </div>
</div>
`;
